### PR TITLE
Feat: Add `TenureChange` transaction type

### LINKED ISSUE #3944: Nakamoto: New `TenureChange` Stacks transaction
In Nakamoto, each bitcoin block may perform a sortition -- much like the Stacks 2.0 protocol performs sortition. However, because the bitcoin commitment no longer contains the current stacks tip (see: #3892), the actual handoff from one miner to the next requires a `TenureChange` stacks transaction.

This is a new `TransactionPayload` variant which indicates that the prior tenure has ended (and what the last Nakamoto block of that tenure was) and that the next tenure should begin.

It should contain:

* The StacksBlockId of the last block from the previous tenure
* The number of blocks produced in the previous tenure
* A flag to indicate which of the following triggered the tenure change
      * A valid winning block-commit
      * No winning block-commits
      * A “null miner” won the block-commit (see the MEV solution below)
* The ECDSA public key hash of the current tenure
* A Schnorr signature from at least 70% of the Stackers
* A bitmap of which Stackers signed

Implementing this transaction payload for mockamoto involves:

1. adding the new transaction variant
2. adding a handler in `db/transactions.rs` which should verify that the current block is a Nakamoto block and that the `StacksBlockId` of the last block from the previous tenure is the current block's parent
3. adding a step in `NakamotoChainState` that checks for the presence of a `TenureChange` transaction, confirming that the current block was signed by the appropriate hash.

### PR DESCRIPTION
**Pull request by jbencin**
### Description

Add a new Stacks transaction type, `TenureChange`, that Stackers will use to control mining tenure in Nakamoto release.

This PR adds the enum variant `TransactionPayload::TenureChange` and establishes the wire format for a tenure change. It does *not* do much in the way of checking whether the trenure change is valid. That will come in a separate PR.

### Applicable issues
- #3944

### Checklist
- [x] Add `TenureChange` variant to `TransactionPayload` enum
- [ ]  In `db/transactions.rs`,  verify that the current block is a Nakamoto block and that the `StacksBlockId` of the last block from the previous tenure is the current block's parent 
- [ ] Add step in `NakamotoChainState` that checks for the presence of a `TenureChange` transaction, confirming that the current block was signed by the appropriate hash
- [ ] Test coverage for new or modified code paths
- [ ] Changelog is updated
- [ ] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [ ] New clarity functions have corresponding PR in `clarity-benchmarking` repo
- [ ] New integration test(s) added to `bitcoin-tests.yml`

### CODE REVIEW

File: stackslib/src/chainstate/stacks/mod.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/mod.rs:660]: CC @xoloki

Comment (xoloki) [reply stackslib/src/chainstate/stacks/mod.rs:660]: A taproot schnorr signature is 64 bytes, using the x-only curve point format.  A regular schnorr proof using secp256k1 would be 65 bytes, with a 33 byte compressed curve point, or 96 bytes with an uncompressed curve point.

Comment (xoloki) [reply stackslib/src/chainstate/stacks/mod.rs:660]: This is the standard `R,s` schnorr proof where `R` is a curve point and `s` is a scalar.

File: stackslib/src/chainstate/stacks/transaction.rs
Comment (jbencin) [reply stackslib/src/chainstate/stacks/transaction.rs:148]: I wouldn't consider these additional exit points, those are created with `return`. The only exit point here is when the output of the `match` expression is returned (unless you count each `?`, in which case there are now fewer)

I can change it back if you want, it's not that important, but IMO it's cleaner and less code to return `Result` types directly when possible, like:

```rust
fn a_function() -> Result<(), Error> {
    // ...
    another_function()
}
```

instead of

```rust
fn a_function() -> Result<(), Error> {
    // ...
    another_function()?;
    Ok(())
}
```

File: stackslib/src/chainstate/stacks/mod.rs
Comment (xoloki) [reply stackslib/src/chainstate/stacks/mod.rs:660]: No reason to allow different formats, I'm just letting y'all know what the options are.  If we're using `wsts` to make a distributed threshold signature, then we get either the 65 or 64 byte formats.  If we're using a libsecp256k1 wrapper for a non distributed threshold signature, then we'll only get the 64 byte format.

Comment (xoloki) [reply stackslib/src/chainstate/stacks/mod.rs:660]: Which we should use depends on what tradeoffs we care about.  If we're going to use `wsts` to make a distributed signature, and also verify it, then the 65 byte format is the cleanest, since it doesn't require jumping through all the hoops necessary for the 64 byte format.  If we care about saving the extra byte, or plan on using a `libsecp256k1` wrapper to verify, then we need to use the 64 byte format.

Comment (jcnelson) [reply stackslib/src/chainstate/stacks/mod.rs:660]: I'm in favor of the 65 byte format in `wsts`.

Comment (xoloki) [reply stackslib/src/chainstate/stacks/mod.rs:660]: Agreed, I think that's the best choice for block signing for `Nakamoto`.  `sBTC` will want to use the 64 byte taproot format, since they need to sign `BTC` tx.

Comment (jbencin) [reply stackslib/src/chainstate/stacks/mod.rs:660]: So a signature in `TenureChange` should consist of both a point and scalar from `wsts`, like this?

```rust
struct SchnorrThresholdSignature {
    pub point: wsts::Point,
    pub scalar: wsts::Scalar,
}
```

File: stackslib/src/chainstate/stacks/transaction.rs
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/transaction.rs:148]: The issue here is that each arm of the `match` expression is now an exit point, so reasoning about where execution returned is a bit harder.  Before, there was only a single exit point.

File: stackslib/src/chainstate/stacks/mod.rs
Comment (xoloki) [reply stackslib/src/chainstate/stacks/mod.rs:660]: Yes, that works perfectly.  If you want a byte representation, you can call

`point.compress().as_bytes()`
`scalar.to_bytes()`

Comment (xoloki) [reply stackslib/src/chainstate/stacks/mod.rs:660]: There's also a `wsts::common::Signature` struct that you can use instead of defining your own.

Comment (jbencin) [reply stackslib/src/chainstate/stacks/mod.rs:660]: Thanks, that's exactly what I need

File: stackslib/src/chainstate/stacks/transaction.rs
Comment (jbencin) [reply stackslib/src/chainstate/stacks/transaction.rs:148]: No problem, I can change it back

File: stackslib/src/chainstate/stacks/db/transactions.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/db/transactions.rs:1364]: A few:

* This must be the first block in the miner's tenure.  It cannot be in a subsequent block
* This `TenureChange` must be for `last_tenure + 1`.  No `TenureChange`s can be skipped.
* It must point to the last active miner's block-commit's `StacksBlockId`
* It must count up the number of blocks since the last block with a `TenureChange` (it may be zero; a block can have multiple `TenureChange` transactions per the design doc)
* It must be signed by the current Stacker set (this could be checked on `consensus_deserialize`)
* It must precede any other kinds of transactions (this could be checked in `StacksBlock::consensus_deserialize`)

File: stackslib/src/chainstate/stacks/mod.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/mod.rs:623]: There is supposed to be a separate `TenureExtend` transaction for this.

File: stackslib/src/chainstate/stacks/mod.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/mod.rs:657]: This is indistinguishable from `SignatureInvalid`.  If less than 70% of signers sign, then they cannot produce a valid signature.

File: stackslib/src/chainstate/stacks/transaction.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/transaction.rs:240]: You're going to somehow need to epoch-gate `consensus_deserialize` below so that this new transaction variant does not even decode until epoch 3.0.  I think we're going to need a separate trait for this (see here: https://github.com/stacks-network/stacks-blockchain/pull/3710#issuecomment-1747137322)

Comment (jbencin) [reply stackslib/src/chainstate/stacks/transaction.rs:240]: Yeah this makes sense. I'll check with Vlad to see if he's working on this for his PR

File: Cargo.lock
Comment (jcnelson) [inline Cargo.lock:2490]: This is a package for RedoxOS, which is not even supported by the blockchain.  Why is this a dependency?

File: stackslib/src/chainstate/stacks/transaction.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/transaction.rs:187]: This is going to require access to a _lot_ more context, such as access to the chainstate databases.  In fact, the code for this might instead belong in `NakamotoChainState`.

File: stackslib/src/chainstate/stacks/transaction.rs
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/transaction.rs:261]: Right now, there are at least two PRs in flight (Vlad's and Cyle's) that do this.  We should just bite the bullet and get one of them shipped, and then base this PR off of it.

Comment (jcnelson) [reply stackslib/src/chainstate/stacks/transaction.rs:261]: Actually, @kantai has already done this in https://github.com/stacks-network/stacks-blockchain/pull/3992.  Can you rebase off of this?

File: Cargo.lock
Comment (jbencin) [reply Cargo.lock:2490]: I added `wsts` and `p256k1` as deps under `./stackslib` but I'm not sure how these got here. I will investigate...

File: stackslib/src/chainstate/stacks/transaction.rs
Comment (fess-v) [reply stackslib/src/chainstate/stacks/transaction.rs:261]: Hey @jbencin @jcnelson ! Can't find in #3992 the logic for `DeserializeWithEpoch` traits for StacksTransaction and StacksBlock, I'm currently working on it and will finish soon, do you think I should wait for that PR to implement this logic and drop mine?

Comment (kantai) [reply stackslib/src/chainstate/stacks/transaction.rs:261]: No -- #3992 isn't going to implement that logic -- so you should continue with your work.

Comment (fess-v) [reply stackslib/src/chainstate/stacks/transaction.rs:261]: Got it, thanks @kantai

File: Cargo.lock
Comment (jbencin) [reply Cargo.lock:2490]: The `Cargo.lock` file on `master` also contains `redox_syscall` and `redox_users`. See [here](https://github.com/stacks-network/stacks-blockchain/blob/e53ba10e75dc26f3dfe92226df2862f1167f806b/Cargo.lock#L2102). Here's the full dependency chain:

```
slog-term -> term -> dirs-next -> dirs-sys-next -> redox_users -> redox_syscall
```

In this PR, whatever I added needs an *additional* version of `redox_syscall` to be specified. This adds the new dependency chain:
```
pk256k1 -> rustfmt-wrapper -> tempfile -> redox_syscall
```

IMO it doesn't matter because as far as I can tell it's never downloaded and built. I don't see it using `cargo tree` or anywhere under `./target`. I guess it's disabled by a feature flag but the dependency gets resolved and pinned anyway?

Comment (jcnelson) [reply Cargo.lock:2490]: Huh, what an odd behavior on Cargo's part.

As long as it's not getting linked into the binary, then I guess this is fine. I wish Cargo would only list packages that actually get built in the `.lock` file.

File: stackslib/src/chainstate/stacks/transaction.rs
Comment (kantai) [reply stackslib/src/chainstate/stacks/transaction.rs:261]: Just to close the loop on this PR: I don't think it's strictly necessary to epoch gate *this* deserialization. All that matters is whether or not the transaction is valid for inclusion in a block. During block processing, the transaction processor should just yield an invalid block error if this transaction payload is matched.

File: stackslib/src/chainstate/stacks/db/transactions.rs
Comment (kantai) [reply stackslib/src/chainstate/stacks/db/transactions.rs:1364]: I think this should live as a separate issue from this PR and let this PR function mostly just as a wire format implementation (https://github.com/stacks-network/stacks-blockchain/issues/4014)

File: stackslib/src/chainstate/stacks/transaction.rs
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/transaction.rs:261]: Yeah that's fine for now.  Furthermore, the block validation code reached from `accept_block()` can reject blocks that include them too early.  I'm already doing this in my coordinator branch.

File: stackslib/src/chainstate/stacks/db/transactions.rs
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/db/transactions.rs:1364]: Agreed

**Comment by jcnelson** [comment]
Before this comes out of draft, we need unit tests for the new functionality.

**Comment by kantai** [comment]
> @jcnelson or @kantai, I'm trying to finish up this PR now, and I've got a couple minor questions regarding how to sign. Right now I have the following struct:
> 
> ```rust
> pub struct TenureChangePayload {
>     pub previous_tenure_end: StacksBlockId,
>     pub previous_tenure_blocks: u16,
>     pub cause: TenureChangeCause,
>     pub pubkey_hash: Hash160,
>     pub signature: ThresholdSignature,
>     pub signers: Vec<u8>,
> }
> ```
> 
> What I'm wondering is:
> 
> * Does the `signature` field need to cover all other fields, including `signers`?

Yes, it needs to cover `signers`.

> * How to do the signature? My guess is that it should be done while the struct is encoded with `StacksMessageCodec` and treat that as a `u8` slice. So then I'd need to pull out `signature` like this:

I would recommend just adding a function:

```rust
impl TenureChangePayload {
    /// Serialize the parts of this payload to `out` required for signing and validating
    /// stacker signatures.
    pub fn serialize_for_sign<W: Write>(&self, out: &mut W) -> Result<(), CodecError> {
          s.previous_tenure_end.consensus_serialize(out)?;
          ...
    }
}
```

**Comment by jbencin** [comment]
> I would recommend just adding a function:

Cool, this works too. The only other change then I'll need is to move `signature` as the last field in the struct, so `serialize_for_sign()` can be called in `TenureChangePayload::consensus_serialize()`.

**Comment by jbencin** [comment]
I've finished up with the structure and wire format for a `TenureChange` transaction. I'm limiting the scope of this PR and pushing most of the validation work off to a later one. With the change in scope this PR is ready for review.

There's one thing that needs to be fixed before merging: There are still 2 unit tests failing. I've been looking into this since yesterday but haven't made any progress:

```
thread 'chainstate::nakamoto::tests::nakamoto_advance_tip_simple' panicked at stackslib/src/chainstate/nakamoto/tests.rs:167:6:
called `Result::unwrap()` on an `Err` value: InvalidStacksBlock("Failed to load total tenure cost from parent. parent_stacks_block_id = 55c9861be5cff984a20ce6d99d4aa65941412889bdc665094136429b84f8c2ee")


failures:
    chainstate::nakamoto::tests::nakamoto_advance_tip_multiple
    chainstate::nakamoto::tests::nakamoto_advance_tip_simple

test result: FAILED. 1241 passed; 2 failed; 125 ignored; 0 measured; 0 filtered out; finished in 416.59s
```

Any ideas what in here might have broken these tests @kantai?

**Comment by jcnelson** [comment]
I've scrapped those two tests in my coordinator PR and replaced them into the `TestPeer`-derived tests (but it's not committed yet).  Going forward, there's going to be a lot of things that these tests will need to do in order to produce validate-able Nakamoto blocks, which `TestPeer` already does for free.

**Comment by kantai** [comment]
> Any ideas what in here might have broken these tests @kantai?

I'm not sure -- I know that I needed to fix those tests when I merged some of this work into my previous PR.

I'd recommend rebasing this PR off of `next` -- that might fix those tests for you. If not, as @jcnelson mentions, those tests are getting supplanted in the coordinator PR anyways, so it may be easier to simply disable them.

**Comment by jbencin** [comment]
> so it may be easier to simply disable them.

Okay, I'm just adding `#[ignore]` to these tests for now

**Comment by kantai** [comment]
Can you rebase this to `next`?

**Comment by jbencin** [comment]
Rebased, force pushed, and changed target to `next`

**Review by kantai — APPROVED** [review]
LGTM!

**Review by jcnelson — APPROVED** [review]
A good chunk of this got merged to `next` as part of the coordinator PR.  If you rebase to `next` then the rest of it is fine with me.  Thanks!

**Comment by jbencin** [comment]
There's one more thing I want to do with `TenureChange`: I want it to use the `wsts::common::Signature` struct so that we can use it's `verify()` method. The only problem is that `Signature` does not implement the necessary derives like `Debug` and `Serialize` (even though `Point` and `Scalar` do), so I can't put it in `TenureChange`

I'll make a PR to WSTS adding the `derive`s, and I'll make a separate PR for this to stacks-core

cc @xoloki

**Comment by jbencin** [comment]
Hey @jcnelson I've resolved the conflicts with `next` and this is ready to merge. Would you like me to wait until you've merged #4074, so we don't risk breaking anything there?

**Pull request merged**
